Register data node stats from info carried back in search responses

dakrone opened the pull request:
This is part of #24915, where we now calculate the EWMA of service time for
tasks in the search threadpool, and send that as well as the current queue size
back to the coordinating node. The coordinating node now tracks this information
for each node in the cluster.

This information will be used in the future the determining the best replica a
search request should be routed to. This change has no user-visible difference.

dakrone commented:
retest this please

s1monw commented on core/src/main/java/org/elasticsearch/action/search/AbstractSearchAsyncAction.java:
```diff
@@ -222,6 +229,11 @@ private void raisePhaseFailure(SearchPhaseExecutionException exception) {
         successfulOps.incrementAndGet();
+        long responseDuration = System.nanoTime() - startNanos;
+        String nodeId = result.getSearchShardTarget().getNodeId();
+        if (nodeId != null && collectorService != null) {
```
why can any of this be null? this looks like a bug?!

s1monw replied:
why do we do this in here and not in the query wrapper. This action might run some DFS stats rather than queries so I think we should capture it solely in the `ResponseListenerWrapper`?

s1monw replied:
if we only capture it in the wrapper we don't need the dependency here no?

s1monw commented on core/src/main/java/org/elasticsearch/action/search/ResponseListenerWrapper.java:
```diff
@@ -0,0 +1,76 @@
+ * result to get the piggybacked queue size and service time EWMA, adding those
+ * values to the coordinating nodes' {@code ResponseCollectorService}.
+ */
+public class ResponseListenerWrapper extends SearchActionListener<SearchPhaseResult> {
```
I think this one can just be an action listener, we can use the `searchShardTarget` from the delegate listener. there is no need to set the targets twice etc.

s1monw replied:
I also wonder if we can find a better name for this

s1monw commented on core/src/main/java/org/elasticsearch/action/search/ResponseListenerWrapper.java:
```diff
@@ -0,0 +1,76 @@
+ */
+public class ResponseListenerWrapper extends SearchActionListener<SearchPhaseResult> {
+
+    private static final Logger logger = ESLoggerFactory.getLogger(ResponseListenerWrapper.class);
```
let's not add logger here. if it's only for debugging we can add it when it's needed

s1monw commented on core/src/main/java/org/elasticsearch/action/search/ResponseListenerWrapper.java:
```diff
@@ -0,0 +1,76 @@
+        if (queryResult != null && collector != null) {
+            long ewma = queryResult.serviceTimeEWMA();
+            int queueSize = queryResult.nodeQueueSize();
+            String nodeId = listener.searchShardTarget.getNodeId();
```
this should never be null. Fail if it is, maybe in the ctor already

s1monw commented on core/src/main/java/org/elasticsearch/action/search/ResponseListenerWrapper.java:
```diff
@@ -0,0 +1,76 @@
+
+    public ResponseListenerWrapper(SearchActionListener<SearchPhaseResult> listener,
+                                   ResponseCollectorService collector) {
+        super(listener.searchShardTarget, listener.requestIndex);
```
maybe we just expect an Actionlistener and a node ID string or even better a DiscoveryNode here. We can get it directly from the connection via `Connection#getNode()` which works with CCS as well then we can reduce visibility again on `SearchActionListener`

s1monw commented on core/src/main/java/org/elasticsearch/action/search/SearchActionListener.java:
```diff
@@ -27,8 +27,9 @@
-    private final int requestIndex;
-    private final SearchShardTarget searchShardTarget;
+
+    protected final int requestIndex;
```
see comment above regarding visibility

s1monw commented on core/src/main/java/org/elasticsearch/node/Node.java:
```diff
@@ -446,6 +447,7 @@ protected Node(final Environment environment, Collection<Class<? extends Plugin>
                 transportService, indicesService, pluginsService, circuitBreakerService, scriptModule.getScriptService(),
                 httpServerTransport, ingestService, clusterService, settingsModule.getSettingsFilter());
             modules.add(b -> {
+                    b.bind(ResponseCollectorService.class).toInstance(responseCollectorService);
```
I don't think we should do this! Let's please don't add any guice related stuff anymore no matter what it costs.

s1monw commented on core/src/main/java/org/elasticsearch/node/ResponseCollectorService.java:
```diff
@@ -0,0 +1,116 @@
+
+    private static final double ALPHA = 0.3;
+
+    private final Map<String, ExponentiallyWeightedMovingAverage> nodeIdToQueueSize;
```
Can we maybe just have a single map and then sync on the node ID with using some KeyedLock? I think that would be simpler?

s1monw commented on core/src/main/java/org/elasticsearch/search/DefaultSearchContext.java:
```diff
@@ -168,7 +174,9 @@
         this.queryResult = new QuerySearchResult(id, shardTarget);
         this.fetchResult = new FetchSearchResult(id, shardTarget);
         this.indexShard = indexShard;
+        this.searchExecutor = indexShard.getThreadPool().executor(ThreadPool.Names.SEARCH);
```
can we do this on the consumer end and specialize the code in SearchService? I think this is where it's used. We shouldn't really add methods to this class it's big enough

dakrone replied:
I don't think I understand, I need a way from `QueryPhase.execute` to access the executor to get the stats out of it, I can't do it on the consumer end since that may be on a different node, and the only thing available in the (static) `QueryPhase.execute` method is the `SearchContext`.

Is there a better place to put this?

s1monw replied:
what I mean is instead of having the `searchExecutor` member can't we do this in `QueryPhase.execute` instead so we don't add more than necessary to the search context. it's already huge and we can access indexShard?

dakrone replied:
That makes sense and is much cleaner, thanks!

s1monw commented on core/src/test/java/org/elasticsearch/node/ResponseCollectorServiceIT.java:
```diff
@@ -0,0 +1,155 @@
+ * Test that response stats are collected for queries executed on a real node
+ */
+@ESIntegTestCase.ClusterScope(scope = ESIntegTestCase.Scope.TEST, numDataNodes = 0)
+public class ResponseCollectorServiceIT extends ESIntegTestCase {
```
instead of doing these old style integration tests should we rather add unittest to do this. I added so many for all the search actions. Let's test all the things in isolation like the wrapper, the query phase etc.

dakrone commented:
retest this please

dakrone commented:
@s1monw thanks for taking a look! I've pushed commits addressing all of your feedback (except for the one I left a question about), could you take another look?

s1monw commented on core/src/main/java/org/elasticsearch/action/search/SearchActionListener.java:
```diff
@@ -27,8 +27,9 @@
+
     private final int requestIndex;
-    private final SearchShardTarget searchShardTarget;
+    protected final SearchShardTarget searchShardTarget;
```
can you add a getter instead of making this protected?

s1monw replied:
given my comment below I think this is unneeded and can be private without a getter

s1monw commented on core/src/main/java/org/elasticsearch/action/search/SearchExecutionStatsCollector.java:
```diff
@@ -0,0 +1,67 @@
+    private final ResponseCollectorService collector;
+    private final long startNanos;
+
+    public SearchExecutionStatsCollector(SearchActionListener<SearchPhaseResult> listener,
```
please instead of exposing the searchShardTarget from `SearchActionListener` can you add a 3rd arg that is the nodeID. we can get it from the passed in connection we talk to instead. then `SearchActionListener` can also be a `ActionListener` instead

s1monw commented on core/src/main/java/org/elasticsearch/action/search/SearchExecutionStatsCollector.java:
```diff
@@ -0,0 +1,67 @@
+ * result to get the piggybacked queue size and service time EWMA, adding those
+ * values to the coordinating nodes' {@code ResponseCollectorService}.
+ */
+public class SearchExecutionStatsCollector implements ActionListener<SearchPhaseResult> {
```
can this be pkg privaate and final?

s1monw commented on core/src/main/java/org/elasticsearch/action/search/SearchTransportService.java:
```diff
@@ -76,10 +76,13 @@
 
-    public SearchTransportService(Settings settings, TransportService transportService) {
+    public SearchTransportService(Settings settings, TransportService transportService,
+                                  ResponseCollectorService responseCollectorService) {
```
you can also remove the dependency by passing a `BiFunction<Connection, SearchActionListener, ActionListener>` instead of ResponseCollectorService and require it to be non-null then it's easier to test? and we never need to do nullchecks

dakrone replied:
I played with this, however, it requires making `SearchExecutionStatsCollector` public instead of package private, since we need to construct it within the `BiFunction` passed in to `SearchTransportService` in `Node`.

I think this is what you had in mind, but let me know if I'm mistaken:

```diff
diff --git a/core/src/main/java/org/elasticsearch/action/search/SearchExecutionStatsCollector.java b/core/src/main/java/org/elasticsearch/action/search/SearchExecutionStatsCollector.java                                                   
index 031e860..96d5e18 100644
--- a/core/src/main/java/org/elasticsearch/action/search/SearchExecutionStatsCollector.java
+++ b/core/src/main/java/org/elasticsearch/action/search/SearchExecutionStatsCollector.java
@@ -23,8 +23,10 @@ import org.elasticsearch.action.ActionListener;
 import org.elasticsearch.node.ResponseCollectorService;
 import org.elasticsearch.search.SearchPhaseResult;
 import org.elasticsearch.search.query.QuerySearchResult;
+import org.elasticsearch.transport.Transport;
 
 import java.util.Objects;
+import java.util.function.BiFunction;
 
 /**
  * A wrapper of search action listeners (search results) that unwraps the query
@@ -47,6 +49,10 @@ final class SearchExecutionStatsCollector implements ActionListener<SearchPhaseR
         this.nodeId = nodeId;
     }
 
+    public static BiFunction<Transport.Connection, SearchActionListener, ActionListener> makeWrapper(ResponseCollectorService service) {                                                                                                    
+        return (connection, originalListener) -> new SearchExecutionStatsCollector(listener, service, connection.getNode().getId());                                                                                                        
+    }
+
     @Override
     public void onResponse(SearchPhaseResult response) {
         QuerySearchResult queryResult = response.queryResult();
diff --git a/core/src/main/java/org/elasticsearch/action/search/SearchTransportService.java b/core/src/main/java/org/elasticsearch/action/search/SearchTransportService.java                                                                 
index 2587623..3ad1517 100644
--- a/core/src/main/java/org/elasticsearch/action/search/SearchTransportService.java
+++ b/core/src/main/java/org/elasticsearch/action/search/SearchTransportService.java
@@ -56,6 +56,7 @@ import org.elasticsearch.transport.TransportResponse;
 import org.elasticsearch.transport.TransportService;
 
 import java.io.IOException;
+import java.util.function.BiFunction;
 import java.util.function.Supplier;
 
 /**
@@ -77,12 +78,13 @@ public class SearchTransportService extends AbstractComponent {
 
     private final TransportService transportService;
     private final ResponseCollectorService responseCollectorService;
+    private final BiFunction<Transport.Connection, SearchActionListener, ActionListener> responseWrapper;
 
     public SearchTransportService(Settings settings, TransportService transportService,
-                                  ResponseCollectorService responseCollectorService) {
+                                  BiFunction<Transport.Connection, SearchActionListener, ActionListener> responseWrapper) {                                                                                                                 
         super(settings);
         this.transportService = transportService;
-        this.responseCollectorService = responseCollectorService;
+        this.responseWrapper = responseWrapper;
     }
 
     public void sendFreeContext(Transport.Connection connection, final long contextId, OriginalIndices originalIndices) {                                                                                                                   
@@ -123,12 +125,7 @@ public class SearchTransportService extends AbstractComponent {
         final boolean fetchDocuments = request.numberOfShards() == 1;
         Supplier<SearchPhaseResult> supplier = fetchDocuments ? QueryFetchSearchResult::new : QuerySearchResult::new;
 
-        final ActionListener handler;
-        if (responseCollectorService == null) {
-            handler = listener;
-        } else {
-            handler = new SearchExecutionStatsCollector(listener, responseCollectorService, connection.getNode().getId());                                                                                                                  
-        }
+        final ActionListener handler = responseWrapper.apply(connection, listener);
         transportService.sendChildRequest(connection, QUERY_ACTION_NAME, request, task,
             new ActionListenerResponseHandler<>(handler, supplier));
     }
diff --git a/core/src/main/java/org/elasticsearch/node/Node.java b/core/src/main/java/org/elasticsearch/node/Node.java
index bb2264f..b10f645 100644
--- a/core/src/main/java/org/elasticsearch/node/Node.java
+++ b/core/src/main/java/org/elasticsearch/node/Node.java
@@ -29,6 +29,7 @@ import org.elasticsearch.ElasticsearchTimeoutException;
 import org.elasticsearch.Version;
 import org.elasticsearch.action.ActionModule;
 import org.elasticsearch.action.GenericAction;
+import org.elasticsearch.action.search.SearchExecutionStatsCollector;
 import org.elasticsearch.action.search.SearchPhaseController;
 import org.elasticsearch.action.search.SearchTransportService;
 import org.elasticsearch.action.support.TransportAction;
@@ -424,8 +425,8 @@ public class Node implements Closeable {
             final TransportService transportService = newTransportService(settings, transport, threadPool,
                 networkModule.getTransportInterceptor(), localNodeFactory, settingsModule.getClusterSettings());
             final ResponseCollectorService responseCollectorService = new ResponseCollectorService(this.settings, clusterService);                                                                                                          
-            final SearchTransportService searchTransportService =  new SearchTransportService(settings,
-                transportService, responseCollectorService);
+            final SearchTransportService searchTransportService =  new SearchTransportService(settings, transportService,                                                                                                                   
+                    SearchExecutionStatsCollector.makeWrapper(responseCollectorService));
             final Consumer<Binder> httpBind;
             final HttpServerTransport httpServerTransport;
             if (networkModule.isHttpEnabled()) {
```

s1monw replied:
yes this looks good to me

s1monw replied:
> I played with this, however, it requires making SearchExecutionStatsCollector public instead of package private, since we need to construct it within the BiFunction passed in to SearchTransportService in Node.

I think that is ok!

s1monw commented on core/src/main/java/org/elasticsearch/node/ResponseCollectorService.java:
```diff
@@ -0,0 +1,144 @@
+ * tasks executed on each node, making the EWMA of the values available to the
+ * coordinating node.
+ */
+public class ResponseCollectorService extends AbstractComponent implements ClusterStateListener {
```
final?

s1monw commented on core/src/main/java/org/elasticsearch/node/ResponseCollectorService.java:
```diff
@@ -0,0 +1,144 @@
+        });
+    }
+
+    public Map<String, ComputedNodeStats> getAllNodeStatistics() {
```
this seems to be test only can we move it into a util in tests?

dakrone replied:
I'm planning on using this method for the very next PR that implements the actual ranking algorithm, without this method the `NodeStatistics` key would be exposed, which is the mutable one. By putting this method here `NodeStatistics` can be private and we can expose only the immutable `ComputedNodeStats`

s1monw replied:
++

s1monw commented on core/src/main/java/org/elasticsearch/node/ResponseCollectorService.java:
```diff
@@ -0,0 +1,144 @@
+     * node's statistics. This includes the EWMA of queue size, response time,
+     * and service time.
+     */
+    public class ComputedNodeStats {
```
can this inner class be static and final and private

s1monw commented on core/src/main/java/org/elasticsearch/node/ResponseCollectorService.java:
```diff
@@ -0,0 +1,144 @@
+     * time, and service time, however, this class is private and intended only
+     * to be used for the internal accounting of {@code ResponseCollectorService}.
+     */
+    private class NodeStatistics {
```
should also be static?

s1monw commented on core/src/main/java/org/elasticsearch/node/ResponseCollectorService.java:
```diff
@@ -0,0 +1,144 @@
+     * to be used for the internal accounting of {@code ResponseCollectorService}.
+     */
+    private class NodeStatistics {
+        public final String nodeId;
```
public is obsolete here

s1monw replied:
serviceTime is non final?! is that expected? just curious...

dakrone commented:
Thanks for the comments @s1monw! I added more changes based on your feedback

dakrone commented:
Thanks for all the reviews @s1monw!
